File: backend/python-fastapi/repositories/user.repository.py
from sqlalchemy.orm import Session
from models.models import User
from uuid import uuid4


class UserRepository:
    @staticmethod
    def find_by_username(db: Session, username: str):
        return db.query(User).filter(User.username == username).first()

    @staticmethod
    def find_by_id(db: Session, id: str):
        return db.query(User).filter(User.id == id).first()

    @staticmethod
    def create(db: Session, username: str, password: str):
        user = User(
            id=str(uuid4()),
            username=username,
            password=password
        )
        db.add(user)
        db.commit()
        db.refresh(user)
        return user

File: backend/laravel/app/Repositories/UserRepository.php
<?php

namespace App\Repositories;

use App\Models\User;

class UserRepository
{
    protected $model;

    public function __construct(User $user)
    {
        $this->model = $user;
    }

    public function findById(int $id): ?User
    {
        return $this->model->find($id);
    }

    public function findByEmail(string $email): ?User
    {
        return $this->model->where('email', $email)->first();
    }

    public function findByUsername(string $username): ?User
    {
        return $this->model->where('username', $username)->first();
    }

    public function create(array $data): User
    {
        return $this->model->create($data);
    }

    public function update(User $user, array $data): bool
    {
        return $user->update($data);
    }

    public function delete(User $user): bool
    {
        return $user->delete();
    }

    public function all(): \Illuminate\Database\Eloquent\Collection
    {
        return $this->model->all();
    }

    public function paginate(int $perPage = 10): \Illuminate\Contracts\Pagination\LengthAwarePaginator
    {
        return $this->model->paginate($perPage);
    }
}
